sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -445,16 +445,25 @@ case class FileSourceScanExec(
       currentSize = 0
     }
 
-    // Assign files to partitions using "Next Fit Decreasing"
-    splitFiles.foreach { file =>
-      if (currentSize + file.length > maxSplitBytes) {
+    def addFile(file: PartitionedFile): Unit = {
+        currentFiles += file
+        currentSize += file.length + openCostInBytes
+    }
+
+    var frontIndex = 0
+    var backIndex = splitFiles.length - 1
+
+    while (frontIndex <= backIndex) {
+        while (frontIndex <= backIndex && currentSize + splitFiles(frontIndex).length <= maxSplitBytes) {
+            addFile(splitFiles(frontIndex))
+            frontIndex += 1
+        }
+        while (backIndex > frontIndex && currentSize + splitFiles(backIndex).length <= maxSplitBytes) {
+            addFile(splitFiles(backIndex))
+            backIndex -= 1
+        }
         closePartition()
-      }
-      // Add the given file to the current partition.
-      currentSize += file.length + openCostInBytes
```

Comment:
this handles files that are larger than the maxSplitBytes, which I think isn't done in the new algorithm.  Need to make sure those don't form an infinite loop

Reply:
should add a test for this too

Reply:
This test still passes https://github.com/glentakahashi/spark/blob/c575977a5952bf50b605be8079c9be1e30f3bd36/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala#L101, and the splitFiles should have been already split by `filesMaxPartitionBytes`

I originally had similar functionality in, but I /think/ its redundant

Reply:
Oh, only if the file `isSplittable`, so yea, will fail for non-splittable files. Will update code

Reply:
saw you added a commit to handle the large non-splittable files case -- can you please add a test for that also?

want to make sure this while loop never becomes an infinite loop!